Fix e2e DNS IPv6 CI

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

> /kind cleanup

> /kind design
> /kind documentation

/kind failing-test

> /kind feature
> /kind flake

**What this PR does / why we need it**:

Fix DNS tests fail with IPv6 https://testgrid.k8s.io/conformance-kind#kind%20(IPv6),%20master%20(dev)

The tests were failing because were trying to use a global variable that's only populated in one of the test executors when running in parallel.

It also make the DNS tests single IP family, because using a probe command asking for A and AAAA records in the command line can lead to false positive, since the command is doing a logical OR instead of a logical AND, that will needed in dual stack clusters.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #81846

**Special notes for your reviewer**:

This simplifies the bash logic to create the pod record reducing the risks

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented:
/cc @BenTheElder 
/hold
we have to verify the IPv6 CI pass before merge, don't know exactly what's causing the problem

johnbelamaric commented:
Do we run the tests with different IPFamily values? Because otherwise in dual stack this will only test one of the sets of records, which I do not think is correct.

johnbelamaric commented:
/priority important-soon

johnbelamaric commented:
/sig network

aojea commented:
> Do we run the tests with different IPFamily values? Because otherwise in dual stack this will only test one of the sets of records, which I do not think is correct.

I guess that dual stack testing is something we should start to brainstorm, my original idea on these DNS tests

https://github.com/kubernetes/kubernetes/pull/75243

 was to use A and AAAA in the same query to make the tests valid for all ip families and dual stack,  but I'm not sure if this is going to work with dual-stack, especially this query

`probeCmd += fmt.Sprintf(`podARec=$$(getent hosts $$(hostname | awk '{print $1}') | tr ":." "-" | awk '{print $$1".%s.pod.%s"}');`, namespace, dnsDomain)`

Since https://github.com/kubernetes/kubernetes/pull/80398 we can detect the IP family of the cluster before the e2e tests start, maybe we can call the functions from the main program with different IP family, avoiding adding more complexity to the function, i.e.

if framework.TextContext.IPFamily = "dual" {
   createProbeCmd = (..., ..., IPv6 true)
   createProbeCmd = (..., ..., IPv6 false)
}

aojea commented:
/test pull-kubernetes-conformance-image-test

aojea commented:
/hold cancel

aojea commented:
/hold

without this fix test pass on my local environment too :man_shrugging:

neolit123 commented:
looks like the problem happened with this PR:
https://github.com/kubernetes/kubernetes/pull/81363
which added separate probes for ipv6, @aojea you also reviewed that PR.

@bclau PTAL

> Let's fix the test failures and tackle dual stack later

+1.

aojea commented:
> looks like the problem happened with this PR:
> #81363
> which added separate probes for ipv6, @aojea you also reviewed that PR.
> 

yeah, but I'm finding inconsistencies on different job runs @neolit123 
Some tests are doing correctly the `if  framework.TestContext.IPFamily == "ipv6"` , however others a fail the comparison, thus use the ipv4 command and fails.

Is it possible that the variable gets override or is not accessible to all tests?

neolit123 commented:
> Is it possible that the variable gets override or is not accessible to all tests?

i'd search the framework for overrides.
you can also run the problem tests locally and see if `framework.TestContext.IPFamily` changes for some reason.

aojea commented:
> > Is it possible that the variable gets override or is not accessible to all tests?
> 
> i'd search the framework for overrides.

> you can also run the problem tests locally and see if `framework.TestContext.IPFamily` changes for some reason.

thanks @neolit123 , seems that some tests are not having that variable, I think that is because of this

https://github.com/kubernetes/kubernetes/blob/d54c5163e041030bd3bc63dcab8876d5f8c51983/test/e2e/network/dns.go#L38

I have to dig more

aojea commented:
/test pull-kubernetes-conformance-kind-ipv6

aojea commented:
/test pull-kubernetes-conformance-kind-ipv6

aojea commented:
/test pull-kubernetes-conformance-kind-ipv6

aojea commented:
/test pull-kubernetes-conformance-kind-ipv6

aojea commented:
/assign @bowei 
/hold cancel

dims commented:
/milestone v1.16

dims commented:
/assign @timothysc 

(conformance related)

neolit123 commented:
AFAIK @timothysc is on PTO this week and this is blocking a lot of pending PRs.
https://testgrid.k8s.io/sig-testing-kind#pull-kubernetes-conformance-kind-ipv6

/assign @spiffxp @oomichi

andrewsykim reviewed (commented):
> thanks @neolit123 , seems that some tests are not passing that variable, I think that is because of this

> I have to dig more, but I can confirm that they are using the probe command for IPv4

What's the updates on this?

aojea commented on test/e2e/framework/suites.go:
```diff
@@ -118,21 +118,29 @@ func SetupSuite() {
+	if err != nil {
+		klog.Fatal("Error loading client: ", err)
+	}
 	TestContext.IPFamily = getDefaultClusterIPFamily(c)
```
it took me a while to discover this, it turns out that this is [how ginkgo run specs in parallel ](https://onsi.github.io/ginkgo/#parallel-specs)

> With SynchronizedBeforeSuite Ginkgo gives you a way to run some preliminary setup code on just one parallel node (Node 1) and other setup code on all nodes. Ginkgo synchronizes these functions and guarantees that node 1 will complete its preliminary setup before the other nodes run their setup code. Moreover, Ginkgo makes it possible for the preliminary setup code on the first node to pass information on to the setup code on the other nodes.

Previously ,TestContext.IPFamily was set on `framework.SetupSuite()`, but it turns out that part of the code only runs on just one parallel node in the first function of `SynchronizedBeforeSuite()`, thus the variable wasn't available for the rest of the nodes.

> SynchronizedBeforeSuite must be passed two functions. The first must return []byte and the second must accept []byte. When running with multiple nodes the first function is only run on node 1. When this function completes, all nodes (including node 1) proceed to run the second function and will receive the data returned by the first function. In this example, we use this data-passing mechanism to forward the database’s address (set up on node 1) to all nodes

I've decided to run the code to the get the variable on the second function from SynchronizedBeforeSuite() because maybe can be useful to other tests in the future, instead of passing the result using the []byte output

```
var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
	// Reference common test to make the import valid.		// Reference common test to make the import valid.
	commontest.CurrentSuite = commontest.E2E		commontest.CurrentSuite = commontest.E2E
	framework.SetupSuite()		framework.SetupSuite()
	return nil		return nil
}, func(data []byte) {	}, func(data []byte) {
	// Run on all Ginkgo nodes		// Run on all Ginkgo nodes
	framework.SetupSuiteAllNodes()
})	})
```

aojea replied:
> So `TestContext.IPFamily` is set _by_ the first parallel node, but does that mean it's not readable from other parallel nodes after the fact? `TestContext` a global var so I assumed it is.

right, actually I was thinking the same when I've implemented it but seems that `TestContext` is global per node executor, not global for all the nodes.

pohly replied:
The first node cannot modify `TestContext` in other nodes because each node runs in a separate process.

`AfterReadingAllFlags` is called by all processes, processing the same input, and thus the result is the same on all nodes. The `getDefaultClusterIPFamily(c)` call could work there, but I think it's better to call it where it is called now because of the need to connect to the API server. Doing that in `AfterReadingAllFlags` "feels" a bit too early.

aojea commented on test/e2e/network/dns_common.go:
```diff
@@ -475,13 +479,16 @@ func createProbeCommand(namesToResolve []string, hostEntries []string, ptrLookup
 
-	if framework.TestContext.IPFamily == "ipv6" {
-		probeCmd += fmt.Sprintf(`podARec=$$(getent hosts $$(hostname | awk '{print $1}') | tr ":." "-" | awk '{print $$1".%s.pod.%s"}');`, namespace, dnsDomain)
+	// getent doesn't work properly on Windows hosts and hostname -i doesn't return an IPv6 address
```
I've tried that approach but with the capital `-I` you obtain *all* the IP addresses, and it overcomplicates the bash oneliner.
```
root@kind-control-plane:/# hostname -i
172.17.0.4
root@kind-control-plane:/# hostname -I
10.244.0.1 10.244.0.1 172.17.0.4 fd00::1 fd00::1 2001:db8:1::242:ac11:4
```
The `getent` command was used previously and worked fine, I tried different things to try to consolidate that line but didn't find a better option, I'm really open to suggestions

aojea commented on test/e2e/framework/suites.go:
```diff
@@ -118,21 +118,29 @@ func SetupSuite() {
+// There are certain operations we only want to run once per overall test invocation on all nodes
+// such as making some global variables accessible to all parallel executions
+// Because of the way Ginkgo runs tests in parallel, we must use SynchronizedBeforeSuite
+func SetupSuiteAllNodes() {
```
done

aojea commented on test/e2e/framework/suites.go:
```diff
@@ -118,21 +118,29 @@ func SetupSuite() {
+}
+
+// SetupSuiteAllNodes is the boilerplate that can be used to setup ginkgo test suites, on the SynchronizedBeforeSuite step.
+// There are certain operations we only want to run once per overall test invocation on all nodes
```
done

aojea commented:
@BenTheElder that test fails just for modifying any file under the `test/e2e/framework` folder :man_shrugging:

aojea committed:
Make TestContext.IPFamily global for parallel testing

it turns out that the framework.TestContext.IPFamily variable is
not available for the DNS tests if they don't run in the initial
Ginkgo node when running in parallel.

We add a function to the framework to allow us to run command
only once per each Ginkgo node parallel execution.

It also adds a method to detect if the cluster is IPv6.

The use of the framework.TestContext.IPFamily variable guarantees
consistency all over the testing because this variable is only
assigned at the beginning of the testing.

aojea committed:
DNS e2e tests differentiate between IP family

The e2e DNS tests differentiate between IP family to
avoid false positives with dual stack clusters.

pohly commented on test/e2e/framework/util.go:
```diff
@@ -428,6 +430,11 @@ func getDefaultClusterIPFamily(c clientset.Interface) string {
+func ClusterIsIPv6() bool {
+	return TestContext.IPFamily == "ipv6"
+}
+
```
One could also argue that this function doesn't belong into `util.go`. As it works on `TestContext`, wouldn't it be better to put this into `test_context.go`? Perhaps even as a method?

```
func (tc TestContextType) ClusterIsIPv6() bool {
    return tc.IPFamily == "ipv6"
}
```

I'm not asking for that change, just wanted to point it out...

aojea replied:
I really think you are right, will submit a PR with that change once this is merged if you don't mind.

aojea replied:
will submit as a follow up once this and your PR disabling the unit test is merged 
https://github.com/kubernetes/kubernetes/commit/45d6f08868ce2729182aae5734a00e5e27ae08f9

pohly replied:
Sounds good. :+1:
